File: exam-conductor/archiveDCR/libs/exampen-proto/python/exampen_proto/plagiarism.py
"""Plagiarism models: flags, evidence, verdicts, check/result events."""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel

from .enums import PlagiarismCheckTrigger, PlagiarismSeverity, TeacherVerdict


class MatchingSegment(BaseModel):
    """Pair of matching text segments between two students."""

    student_a_text: str
    student_b_text: str


class Evidence(BaseModel):
    """Plagiarism detection evidence for a flag."""

    matching_segments: list[MatchingSegment]
    temporal_correlation_score: Optional[float] = None
    seating_proximity_score: Optional[float] = None


class FlagSummary(BaseModel):
    """Lightweight plagiarism flag for list views."""

    flag_id: UUID
    exam_id: UUID
    student_a_id: str
    student_b_id: str
    question_id: str
    composite_score: float
    severity: PlagiarismSeverity
    teacher_verdict: Optional[TeacherVerdict] = None


class FlagDetail(FlagSummary):
    """Full plagiarism flag with evidence and verdict detail."""

    evidence: Evidence
    verdict_reason: Optional[str] = None
    verdict_by: Optional[str] = None
    verdict_at: Optional[datetime] = None


class VerdictRequest(BaseModel):
    """Teacher verdict on a plagiarism flag."""

    teacher_id: str
    verdict: TeacherVerdict
    reason: str


class PlagiarismFlagEvent(BaseModel):
    """Single flag entry within a plagiarism.result event."""

    flag_id: UUID
    student_a_id: str
    student_b_id: str
    question_id: str
    composite_score: float
    severity: PlagiarismSeverity


class PlagiarismCheckEvent(BaseModel):
    """NATS event: plagiarism check requested."""

    event_id: str
    event_type: str = "plagiarism.check"
    event_version: str = "1.0.0"
    occurred_at: datetime
    exam_id: UUID
    student_count: int
    question_count: int
    trigger: PlagiarismCheckTrigger


class PlagiarismResultEvent(BaseModel):
    """NATS event: plagiarism check completed with flags."""

    event_id: str
    event_type: str = "plagiarism.result"
    event_version: str = "1.0.0"
    occurred_at: datetime
    exam_id: UUID
    flags: list[PlagiarismFlagEvent]

File: exam-conductor/archiveDCR/libs/exampen-proto/typescript/src/plagiarism.ts
/** Plagiarism types: flags, evidence, verdicts, check/result events. */

import type {
  PlagiarismCheckTrigger,
  PlagiarismSeverity,
  TeacherVerdict,
} from "./enums";

/** Pair of matching text segments between two students. */
export interface MatchingSegment {
  student_a_text: string;
  student_b_text: string;
}

/** Plagiarism detection evidence for a flag. */
export interface Evidence {
  matching_segments: MatchingSegment[];
  temporal_correlation_score?: number;
  seating_proximity_score?: number;
}

/** Lightweight plagiarism flag for list views. */
export interface FlagSummary {
  flag_id: string;
  exam_id: string;
  student_a_id: string;
  student_b_id: string;
  question_id: string;
  composite_score: number;
  severity: PlagiarismSeverity;
  teacher_verdict?: TeacherVerdict;
}

/** Full plagiarism flag with evidence and verdict detail. */
export interface FlagDetail extends FlagSummary {
  evidence: Evidence;
  verdict_reason?: string;
  verdict_by?: string;
  verdict_at?: string;
}

/** Teacher verdict on a plagiarism flag. */
export interface VerdictRequest {
  teacher_id: string;
  verdict: TeacherVerdict;
  reason: string;
}

/** Single flag entry within a plagiarism.result event. */
export interface PlagiarismFlagEvent {
  flag_id: string;
  student_a_id: string;
  student_b_id: string;
  question_id: string;
  composite_score: number;
  severity: PlagiarismSeverity;
}

/** NATS event: plagiarism check requested. */
export interface PlagiarismCheckEvent {
  event_id: string;
  event_type: "plagiarism.check";
  event_version: "1.0.0";
  occurred_at: string;
  exam_id: string;
  student_count: number;
  question_count: number;
  trigger: PlagiarismCheckTrigger;
}

/** NATS event: plagiarism check completed with flags. */
export interface PlagiarismResultEvent {
  event_id: string;
  event_type: "plagiarism.result";
  event_version: "1.0.0";
  occurred_at: string;
  exam_id: string;
  flags: PlagiarismFlagEvent[];
}
